### PR TITLE
mgr/cephadm: pass daemon's current image when reconfiguring

### PR DESCRIPTION
**Pull request by adk3798**
Important to note here is that a reconfig will rewrite the config files
of the daemon and restart it, but not rewrite the unit files. This lead
to a bug where the grafana image we used between the quincy and
squid release used different UIDs internally, which caused us to rewrite
the config files as owned by a UID that worked with the new image but
did not work with the image still specified in the unit.run file. This
meant the grafana daemon was down from a bit after the mgr upgrade
until the end of the upgrade when we redeploy all the monitoring images.

Fixes: https://tracker.ceph.com/issues/65234

---

This is implicitly tested by some of the unit tests and the failed teuthology test referenced in the tracker ticket.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by ljflores — APPROVED** [review]

File: src/pybind/mgr/cephadm/module.py [hunk 1]
```diff
@@ -1582,9 +1582,27 @@ def _client_keyring_rm(
         self._kick_serve_loop()
         return HandleCommandResult()
 
-    def _get_container_image(self, daemon_name: str) -> Optional[str]:
+    def _get_container_image(self, daemon_name: str, use_current_daemon_image: bool = False) -> Optional[str]:
```
Comment (ljflores) [inline src/pybind/mgr/cephadm/module.py:1585]: The failure from run-tox-mgr looks like it has a problem with this new parameter:

```
Expected: _run_cephadm('hostx', 'smb.tango.briskly', ['_orch', 'deploy'], [], stdin='{"fsid": "fsid", "name": "smb.tango.briskly", "image": "", "deploy_arguments": [], "params": {}, "meta": {"service_name": "smb", "ports": [], "ip": null, "deployed_by": [], "rank": null, "rank_generation": null, "extra_container_args": null, "extra_entrypoint_args": null}, "config_blobs": {"cluster_id": "foxtrot", "features": ["domain"], "config_uri": "rados://.smb/foxtrot/config2.json", "join_sources": ["rados://.smb/foxtrot/join1.json", "rados:mon-config-key:smb/config/foxtrot/join2.json"], "custom_dns": ["10.8.88.103"], "config": "", "keyring": "[client.smb.config.tango.briskly]\\nkey = None\\n\\n[client.smb.fs.cephfs.share1]\\nkey = None\\n\\n[client.smb.fs.cephfs.share2]\\nkey = None\\n\\n[client.smb.fs.fs2.share3]\\nkey = None\\n", "config_auth_entity": "client.smb.config.tango.briskly"}}')
E           Actual: _run_cephadm('hostx', 'smb.tango.briskly', ['_orch', 'deploy'], [], stdin='{"fsid": "fsid", "name": "smb.tango.briskly", "image": "", "deploy_arguments": [], "params": {}, "meta": {"service_name": "smb", "ports": [], "ip": null, "deployed_by": [], "rank": null, "rank_generation": null, "extra_container_args": null, "extra_entrypoint_args": null}, "config_blobs": {"cluster_id": "foxtrot", "features": ["domain"], "config_uri": "rados://.smb/foxtrot/config2.json", "join_sources": ["rados://.smb/foxtrot/join1.json", "rados:mon-config-key:smb/config/foxtrot/join2.json"], "custom_dns": ["10.8.88.103"], "config": "", "keyring": "[client.smb.config.tango.briskly]\\nkey = None\\n\\n[client.smb.fs.cephfs.share1]\\nkey = None\\n\\n[client.smb.fs.cephfs.share2]\\nkey = None\\n\\n[client.smb.fs.fs2.share3]\\nkey = None\\n", "config_auth_entity": "client.smb.config.tango.briskly"}}', use_current_daemon_image=False)

/usr/lib/python3.10/unittest/mock.py:929: AssertionError
```

Comment (ljflores) [reply src/pybind/mgr/cephadm/module.py:1585]: Expected vs. actual is the same except for:

```
use_current_daemon_image=False
```

Comment (adk3798) [reply src/pybind/mgr/cephadm/module.py:1585]: I think I must have written this patch before rebasing on top of the recent SMB support patches in cephadm. Hopefully good now.

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2024-04-08_12:15:10-orch:cephadm-wip-adk-testing-2024-04-04-1735-distro-default-smithi/

failures were all either `in cluster log` stuff or mds_upgrade_sequence (both known issues) except one instance of https://tracker.ceph.com/issues/65155 because I forgot to tag the PR that fixes it.

Nothing to block merging

**Pull request merged**